dvc/repo/plots/__init__.py

```diff
@@ -270,19 +270,35 @@ def _is_plot(out: "Output") -> bool:
 
 
 def _resolve_data_sources(plots_data: Dict):
-    for value in plots_data.values():
+    values = list(plots_data.values())
+    to_resolve = []
+    while values:
+        value = values.pop()
         if isinstance(value, dict):
             if "data_source" in value:
-                data_source = value.pop("data_source")
-                assert callable(data_source)
-                value.update(data_source())
-            _resolve_data_sources(value)
+                to_resolve.append(value)
+            values.extend(value.values())
+
+    def resolve(value):
+        data_source = value.pop("data_source")
+        assert callable(data_source)
+        value.update(data_source())
+
+    if len(to_resolve) > 1:
```

Comment:
@skshetry do you know what it the overhead creating a TP? It feels it can up to a second? I put some optimizations to avoid that ... I wonder of we can detect that files are in cache already on all of them are local to avoid creating it at all